src/prefect_server/api/flow_concurrency_limits.py

```diff
@@ -0,0 +1,162 @@
+    """
+
+    if limit <= 0:
+        raise ValueError(
+            (
+                "Flow concurrency limits require positive integers representing"
+                f" the maximum number of concurrent allowed flows. Got {limit}"
+            )
+        )
+
+    existing = await models.FlowConcurrencyLimit.where(
+        {"name": {"_eq": name}, "tenant_id": {"_eq": tenant_id}}
+    ).first({"id", "name", "limit"})
+
+    if not existing:
+        return await models.FlowConcurrencyLimit(
+            name=name, limit=limit, tenant_id=tenant_id
+        ).insert()
+
+    if not existing.limit == limit:
+        result = await models.FlowConcurrencyLimit.where(id=existing.id).update(
+            set=dict(limit=limit)
+        )
+        if not bool(result.affected_rows):
+            raise ValueError("Error updating existing flow concurrency limit")
+
+    return existing.id
+
+
+@register_api("flow_concurrency_limits.delete_flow_concurrency_limit")
+async def delete_flow_concurrency_limit(limit_id: str) -> bool:
+    """
+    Deletes a flow concurrency limit.
+
+    Args:
+        - limit_id (str): The flow concurrency limit to delete.
+
+    Returns:
+        - bool: If the delete was successful
+
+    Raises:
+        - ValueError: If an ID isn't provided
+    """
+
+    if not limit_id:
+        raise ValueError("Invalid flow concurrency limit ID.")
+
+    result = await models.FlowConcurrencyLimit.where(id=limit_id).delete()
+    return bool(result.affected_rows)
+
+
+@register_api("flow_concurrency_limits.try_take_flow_concurrency_slots")
+async def try_take_flow_concurrency_slots(
+    tenant_id: str, limit_names: List[str], current_state: Optional[State] = None
+) -> bool:
+    """
+    Determines whether a `flow_run_id` either already occupies
+    a concurrency slot or can occupy a concurrency slot. If
+    a `flow_run_id` is not provided, the `limit_names` are searched
+    in general, seeing if there are > 0 available sots.
+
+    A concurrency slot is "already occupied" if the flow run
+    is already in a Submitted State or Running State, but
+    there isn't a point of trying to grab a concurrency
+    slot if the flow run is already running, so the more common
+    transition is from Submitted -> Running.
+
+    Args:
+        - tenant_id (str): Tenant owning the flow_run
+        - limit_names (List[str]): Concurrency limits that may
+            or may not exist. Nonexistant limits are treated
+            as unlimited.
+        - current_state (Optional[State]): Current state of the
+            flow run attempting to be allocated a slot.
+            If not provided, skips that check.
+
+    Returns:
+        - bool: Whether the run already occupies or can occupy a concurrency
+            slot.
+    """
+
+    if not limit_names:
+        # Unlabeled runs always have available slots
+        return True
+
+    if current_state is not None:
+        is_occupying_run_slot = (
```

Comment:
Technically we only need to check if the state is `Submitted` here, since flow runs in a `Running` state won't ever need to acquire a concurrency slot. _However_, I think leaving the check for `Running` and `Submitted` helps keep the intention clear since we're checking `Running` and `Submitted` later in the function.